cmake: package_helper CMake script

tejlmand opened the pull request:
The Zephyr CMake package helper script allows to invoke Zephyr package
modules as they are invoked at configure time, up to the modules given
with `-DMODULES=<modules>`.
This allows to run the configure time part of the Zephyr build system
without generating a complete build system.

This means that for example devicetree module can be invoked for a given
sample and board to generate zephyr.dts identical to what is created
by the build system during a build.

To generate zephyr.dts for hello_world, invoke the script as:
$ cmake -DBOARD=<board> -B build -S samples/hello_world \
$       -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_helper.cmake

Other modules, for example 'kconfig' and 'zephyr_modules' can also be
invoked this way.

This will allow other tools, such as twister, to executed only a
subset of the build system like dts in order to do filtering before
determine if a complete build should be invoked.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

---------

**NOTE** Ignore first commit, it's a squash of #41301.

tejlmand committed:
cmake: package_helper CMake script

The Zephyr CMake package helper script allows to invoke Zephyr package
modules as they are invoked at configure time, up to the modules given
with `-DMODULES=<modules>`.
This allows to run the configure time part of the Zephyr build system
without generating a complete build system.

This means that for example devicetree module can be invoked for a given
sample and board to generate zephyr.dts identical to what is created
by the build system during a build.

To generate zephyr.dts for hello_world, invoke the script as:
$ cmake -DBOARD=<board> -B build -S samples/hello_world \
$       -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_helper.cmake

Other modules, for example 'kconfig' and 'zephyr_modules' can also be
invoked this way.

This will allow other tools, such as twister, to executed only a
subset of the build system like dts in order to do filtering before
determine if a complete build should be invoked.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,113 @@
+#
+#   you can invoke only dts module (and dependencies) as:
+#   $ cd samples/hello_world/build
+#   $ cmake -DBOARD=<board> -DAPPLICATION_SOURCE_DIR=.. \
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
```
This correctly generated zephyr/edt.pickle in the build directory (without having to run the full build) for me. That pickle file should be the only thing that twister really needs, so I  this PR is a step towards solving the problem of being able to do devicetree filtering before running a full CMake for a basic build.

However, I can't tell if it works with `SHIELD`, since the rework to use cmake modules seems to have broken the printing of that in BOARD_MESSAGE.

I assume you meant it to work for all build settings due to the comment below.

tejlmand replied:
> However, I can't tell if it works with SHIELD, since the rework to use cmake modules seems to have broken the printing of that in BOARD_MESSAGE.

This should be fixed now.
This PR has been rebased against #41301 to get the latest review comments and fixes such as shield printing.

mbolivar-nordic commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,113 @@
+#           -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
+#
+# Note: the samples CMakeLists.txt file is not processed by package tooling, so
+#       any 'set(<var> <value>)' specified before 'find_package(Zephyr)' must be
+#       manually applied, for example if the CMakeLists.txt contains:
```
This is an unfortunate limitation, but it makes sense why it has to be this way.

tejlmand replied:
I have an idea that I will test out which might remove this limitation.

tejlmand commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,113 @@
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# The Zephyr package tooling script provides a generic script mode interface
```
Thanks for the proposal.
I'm not completely fond of `package_tooling.cmake`, and maybe `run_cmake_module.cmake` is better.
Downside with that name is that it can give the impression that it's possible to run any CMake module, when in fact it only allows to run Zephyr cmake modules (and it does so through the `find_package(Zephyr)` call).

Before changing, I would like to see of other good proposals arises.

tejlmand replied:
> Maybe run_cmake_module.cmake would be a better name for this file?

How about calling it `breeze.cmake` ?

Then `west build` can do a full Zephyr or just a light breeze in future :wink:

mbolivar-nordic replied:
> How about calling it breeze.cmake ?

Discussed offline with @carlescufi - both of us think that the name should make it obvious what this file does instead, so `breeze` doesn't fit the bill

mbolivar-nordic replied:
How about `package_helper`?

tejlmand replied:
`package_helper` is fine with me.

mbolivar-nordic replied:
Care to rename this to package_helper, then? I am happy to approve after that.

tejlmand replied:
done.

Was just being a bit passive on this part, as I'm anyway waiting for #41301 to go in first.

tejlmand commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,113 @@
+#   $ cmake -DBOARD=<board> ..
+#
+#   you can invoke only dts module (and dependencies) as:
+#   $ cd samples/hello_world/build
```
> I think this would be clearer if you added the line mkdir samples/hello_world/build before it.

Good point.
Will do.

>  these comments made me think I had to run the full cmake build system generation step before I could run this, which would defeat the purpose of the PR.

Actually my long term plan is to get this integrated with `west build` as an option, like:
`west build --only-dts -b<board> <sample>` or perhaps `west build --cmake-module=dts -b<board> <sample>`.
With such an approach, a  user can use the normal build flow, but ask the build system to process up to and including a specific module.

But I intent to do that separately from this PR.

tejlmand replied:
> > I think this would be clearer if you added the line mkdir samples/hello_world/build before it.
>
> Good point.
> Will do.

Done

tejlmand replied:
@mbolivar-nordic with latest support of `-B <path-to-build>` then the whole `mkdir build` and `cd build` can be removed.
This is more in line as what can be done directly with CMake.

Of course `mkdir` and `cd` is still possible, but we don't need to describe all various ways CMake can be invoked.

tejlmand commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,113 @@
+#   $ cd samples/hello_world/build
+#   $ cmake -DBOARD=<board> ..
+#
+#   you can invoke only dts module (and dependencies) as:
```
fixed

SebastianBoe commented:
Why doesn't -B and -H work?

Can it be supported?

sebo@sebo:~/ncs/zephyr$ cmake -DBOARD=bbc_microbit -DAPPLICATION_SOURCE_DIR=samples/hello_world -DMODULES=boards -P cmake/package_tooling.cmake -Boe -H.
Loading Zephyr module(s) (Zephyr base): python;west;root;zephyr_module;boards;shields;build_configuration;boards
-- Found Python3: /usr/bin/python3.9 (found suitable exact version "3.9.5") found components: Interpreter 
-- Found west (found suitable version "0.12.0", minimum required is "0.7.1")
-- CMAKE_BINARY_DIR: /home/sebo/ncs/zephyr
CMake Error at cmake/modules/zephyr_module.cmake:63 (file):
  file problem creating directory: /home/sebo/ncs/zephyr/Kconfig
Call Stack (most recent call first):
  share/zephyr-package/cmake/ZephyrConfig.cmake:49 (include)
  share/zephyr-package/cmake/ZephyrConfig.cmake:62 (include_boilerplate)
  cmake/package_tooling.cmake:113 (find_package)

SebastianBoe commented on cmake/modules/boards.cmake:
```diff
@@ -154,4 +154,6 @@ if(NOT BOARD_DIR)
-add_custom_target(boards ${list_boards_commands} USES_TERMINAL)
+if(NOT CMAKE_SCRIPT_MODE_FILE)
+  add_custom_target(boards ${list_boards_commands} USES_TERMINAL)
+endif()
```
Should we document the desired outcome when boards.cmake is executed in script mode?

And perhaps mention in a comment in this file that it is supported to do so?

Otherwise we might break the intended behaviour by accident.

tejlmand replied:
It should.

Long term, then we could even consider to remove support for `ninja boards`.

After all, user has successfully generated a build system, so what's the real value in having a dedicated target that can list all boards ?

@mbolivar-nordic what do you think, is it time to remove `ninja boards` and go purely for the python scripts ?
This should probably go hand-in-hand with a `west shields` / `scripts/list_shields.py` script.

tejlmand replied:
@SebastianBoe I decided to override the `add_custom_target()` / `set_target_properties()` in the script to silence the CMake errors.

This removes the need for special handling / testing for `CMAKE_SCRIPT_MODE_FILE` at multiple locations in Zephyr CMake modules because the new approach is localized to the tooling script.
It also makes the commenting clearer.
https://github.com/zephyrproject-rtos/zephyr/blob/184940230028729634828431073444cd97733874/cmake/package_tooling.cmake#L50-L62

this allows us to write Zephyr CMake modules as we would normally do, and do the special handling in the code that wants to run only a subset of the build system.

SebastianBoe reviewed (changes requested):
changes requested

tejlmand commented:
> Why doesn't -B and -H work?

Let me see what can be done wrt. `-B`.

For `-H`, that means `--help` (changed years ago).
You probably mean `-S` now.
https://cmake.org/cmake/help/latest/manual/cmake.1.html
> -S <path-to-source>
>    Path to root directory of the CMake project to build.

> --help,-help,-usage,-h,-H,/?
>    Print usage information and exit.
>    Usage describes the basic command line interface and its options.

will see what can be done wrt. `-S` also.

tejlmand commented:
and the reason they don't work currently, is because of script mode.

tejlmand commented:
> Why doesn't -B and -H work?
>
> Can it be supported?

Fixed, now `-S` and `-B` is supported.
`APPLICATION_BINARY_DIR` has been removed as it is no longer needed (because we now support `-B`).

`-S` takes the source dir, and sets `APPLICATION_SOURCE_DIR` before calling `find_package(Zephyr)`.

This means running this tool can be done with common CMake arguments, like:
```
$ cmake -DBOARD=bbc_microbit -S samples/hello_world -DMODULES=dts -P cmake/package_tooling.cmake -Bbuild
```

SebastianBoe commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,128 @@
+# dts module is loaded in Kconfig mode.
+if(kconfig IN_LIST MODULES)
+  list(APPEND BASE_MODULES dts)
+endif()
```
Why doesn't dts include Kconfig?

tejlmand replied:
Because `dts` is run before `kconfig`, and hence, to run only the `dts` for Zephyr you don't need to run `kconfig`.

But if you want to run `kconfig` as is done in a regular Zephyr build, then you must run `dts` in order to have the devicetree knowledge available inside `kconfig` tree.

SebastianBoe replied:
Sorry, I wrote the dependency the wrong way. Can you answer again for the correct dependency direction?

tejlmand replied:
because the `kconfig` module itself doesn't depend on `dts`, but the way Zephyr Kconfig tree is written means that when Kconfig module is run in a Zephyr build, then `dts` is required.

This small difference becomes clearer when also looking at [SYSBuild](https://github.com/zephyrproject-rtos/zephyr/pull/40555), where SYSbuild re-uses Kconfig for it's menu, but doesn't re-use `dts`.
https://github.com/zephyrproject-rtos/zephyr/blob/e31348329a5e04f8c9293ac50bdb9918d2a238fe/sysbuild/CMakeLists.txt#L15-L16

This shows that Kconfig itself can be used without `dts`, but when wanting to run kconfig for a Zephyr build, then dts is required.
This is why this knowledge exists here, and not inside the kconfig CMake module.
Just like dts is loaded before kconfig here:
https://github.com/zephyrproject-rtos/zephyr/blob/184940230028729634828431073444cd97733874/cmake/modules/zephyr_default.cmake#L85-L86

SebastianBoe replied:
It would be unfortunate if we were overly generic for a future feature, and then that future feature doesn't get merged.

I had to revert lots of changes in master for the multi-image build because they were overly generic and then the
multi-image build was denied.

Can we do include(dts) now, and then, when we need to support non-DT invocation of Kconfig we make Kconfig agnostic to DT?

tejlmand replied:
> It would be unfortunate if we were overly generic for a future feature, and then that future feature doesn't get merged.

I know, that's why #40555 and the strategy behind it has already been presented both at the dev-review and the TSC, with very positive feedback.

And exactly due to that feedback and some existing additional use-cases, work like #41301 was started.

SebastianBoe replied:
Today, kconfig always uses DT, so it should include it as a dependency.

The future users that intend to decouple this dependency can do so in that PR.

tejlmand replied:
> Today, kconfig always uses DT, so it should include it as a dependency.

No, cause the Kconfig CMake module itself doesn't depend on dts.
The Kconfig tree does.

Just like the Kconfig CMake module doesn't depends on Zephyr modules, but our Kconfig tree uses it.

This is why there we have this:
https://github.com/zephyrproject-rtos/zephyr/blob/3df554fcb2c15c09278477bb4e04888d44683b19/cmake/modules/zephyr_default.cmake#L85-L86

That file is the order in which the Zephyr build itself expects things to be executed, but that is not the same as the Zephyr CMake modules themselves has such dependency.

We shouldn't artificial dependencies between modules, just because that's how Zephyr build system expects to execute them.

SebastianBoe replied:
A suggestion for getting rid of this dependency information has been proposed here[0], making the question of if kconfig should include dts moot.

[0] https://github.com/zephyrproject-rtos/zephyr/pull/41302/files#r797716740

SebastianBoe commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,128 @@
+if(NOT DEFINED MODULES)
+  message(WARNING
+    "No MODULES defined, please invoke tooling with minimum one module"
+    " to execute in script mode."
```
Could we list the scripts that are supported to run in script mode here?

Or are all scripts supported?

I'm concerned users will generate support questions like "can I run script X" in script mode?

tejlmand replied:
the intention with this script is to be able to run the Zephyr build system as is done in configure time (not generator time), and stop at any point before the `kernel.cmake` module.

That will then produce the configure time output from that module, as if it had been executed as part of Zephyr CMake `configure` time. This allows to examine the output.

Such output could be `.config`, `zephyr.dts`, `zephyr_modules.txt`, etc.

Especially useful for twister, which allows to do `dts` or `kconfig` and then filter the output to see if a given sample + board meets a pre-condition for running a full build.

SebastianBoe replied:
So skipping scripts that are not needed for the desired target is out-of-scope?

tejlmand replied:
> So skipping scripts that are not needed for the desired target is out-of-scope?

not sure I understand the question.
Can you elaborate.

Did you read the purpose of this PR: https://github.com/zephyrproject-rtos/zephyr/pull/41302#issue-1083249333
> The Zephyr CMake package tooling script allows to invoke individual
Zephyr package modules (and their dependencies) in script mode without
running the complete CMake configure and build system generation stages.
>
> This means that for example devicetree module can be invoked for a given
sample and board to generate zephyr.dts identical to what is created
by the build system during a build.

and:
> This will allow other tooling, such as twister, to executed only a
subset of the build system like dts in order to do filtering before
determine if a complete build should be invoked.

SebastianBoe replied:
There are two strategies, either

1. run the exact same scripts in the exact same order and then stop when we have achieved the desired result.

2. run the desired target and rely on it including it's dependencies.

The PR description is describing 2. But this sentence is describing 2:

"the intention with this script is to be able to run the Zephyr build system as is done in configure time (not generator time), and stop at any point before the kernel.cmake module."

I'm wondering if the minimal execution that we get with 2. is in-scope, or if it is the simpler strategy 1. we are going for.

tejlmand replied:
ok, maybe `dependencies` in the PR description is misleading.

It's the execution order / dependencies seen from the regular Zephyr CMake configure time, not the internal Zephyr CMake module dependencies.

So the correct answer is 1.

tejlmand replied:
I have updated the description, and the code to make this more clear.

SebastianBoe replied:
This is now resolved. Execution can stop after any script in zephyr_default.cmake is executed.

SebastianBoe commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,128 @@
+
+# List of base modules needed in order to ensure same behavior as the default
+# Zephyr build system during configure time.
+set(BASE_MODULES python west root zephyr_module boards shields arch configuration_files)
```
Dependencies between our internal CMake scripts used to be encoded in one place. boilerplate.cmake.

Now it's encoded in three places. Each individual cmake script, zephyr_default.cmake, and package_tooling.cmake.

Can we get it down to two places for maintainability reasons?

I'm concerned that when a new module is added people will forget to add it to all three locations.

e.g. we could have package_tooling.cmake execute zephyr_default.cmake and have zephyr_default.cmake return when we had executed enough code. So in zephyr_default.cmake it would look like:

```
include(dts)
if(STOP_AFTER_DTS)
return()
endif()

include(kconfig)
if(STOP_AFTER_KCONFIG)
return()
endif()

```

SebastianBoe replied:
I see it's been re-organized.

But could we have both zephyr_default and package_tooling re-use this same list ?

tejlmand replied:
> Dependencies between our internal CMake scripts used to be encoded in one place. boilerplate.cmake.

and this corresponds to `zephyr_default.cmake` if you want to look at it from that perspective.

Now, you could argue that `zephyr_default.cmake` doesn't need to include Zephyr CMake modules if they will anyway be loaded by their dependencies, but that will make it harder to lookup at a single location to see what gets loaded.

So that's why I prefer to list all CMake modules in `zephyr_default.cmake`, so that it's visible when it gets loaded.

> Now it's encoded in three places. 
> Each individual cmake script,

Because those are modules intended for re-usability, so they might have additional strong dependencies.
This is what I have been trying to explain here: 
https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777444109
https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777553129

and as said in that PR, it's a first step in the process.

>  zephyr_default.cmake, and package_tooling.cmake.
> To reduce the number of places we list dependencies we could drop this list and instead have package_tooling.cmake execute zephyr_default.cmake and have zephyr_default.cmake return when we had executed enough code.

Now, this is where I agree it should be possible to improve, as to ensure `zephyr_default.cmake` and `package_tooling.cmake` is using a synced list.

Will take a look at this.

tejlmand replied:
Pushed a fix.

`find_package(Zephyr COMPONENTS <component>)` now supports indicating subparts to load.

This allows for constructs like this:
`find_package(Zephyr COMPONENTS zephyr_default:dts)`

which will load `zephyr_default` until the module `dts` is loaded.

This means `package_tooling` no longer need to have any knowledge on this.
It simply passes on the user arguments to find_package, and then let's zephyr_default do the magic.

tejlmand committed:
cmake: support sub-component loading in zephyr_default module

This commit allows to load only subcomponents of the default Zephyr
CMake module, zephyr_default.cmake.

This allows other tools to execute Zephyr CMake build system up to a
specific module and the stop further processing.

This commit is an enabler for future support in twister to process only
devicetree or kconfig, to allow test / sample filtering before
generating a complete build system.

Sub-components can be loaded as:
> find_package(Zephyr COMPONENTS zephyr_default:<sub-component>)

for example:
> find_package(Zephyr COMPONENTS zephyr_default:dts)

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic commented on cmake/modules/zephyr_default.cmake:
```diff
@@ -32,41 +32,42 @@ find_package(ZephyrBuildConfiguration
 
 # Include board specific device-tree flags before parsing.
-include(${BOARD_DIR}/pre_dt_board.cmake OPTIONAL)
+set(pre_dt_board "\${BOARD_DIR}/pre_dt_board.cmake" OPTIONAL)
+list(APPEND sub_components "\${pre_dt_board}")
```
I don't follow how this works with the loop below. How are you handling the OPTIONAL part?

tejlmand replied:
This is handled here:
https://github.com/zephyrproject-rtos/zephyr/blob/8f2599cdb1284dba80532a2c7cf580ffb884b796/cmake/modules/zephyr_default.cmake#L117-L118

First time, the `${pre_dt_board}` from here: https://github.com/zephyrproject-rtos/zephyr/blob/8f2599cdb1284dba80532a2c7cf580ffb884b796/cmake/modules/zephyr_default.cmake#L70
gets expanded, second time the `\${BOARD_DIR}` from here https://github.com/zephyrproject-rtos/zephyr/blob/8f2599cdb1284dba80532a2c7cf580ffb884b796/cmake/modules/zephyr_default.cmake#L69
gets expanded.

Which results in the following code being executed. (CMake trace output)
```
.../zephyr/cmake/modules/zephyr_default.cmake(117):  string(CONFIGURE ${pre_dt_board} module )
.../zephyr/cmake/modules/zephyr_default.cmake(118):  string(CONFIGURE ${BOARD_DIR}/pre_dt_board.cmake;OPTIONAL module )
.../zephyr/cmake/modules/zephyr_default.cmake(119):  include(/projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/pre_dt_board.cmake;OPTIONAL )
```

so as you see, the `OPTIONAL` is expanded as expected and passed to the `include()` call as argument.

SebastianBoe commented on cmake/modules/zephyr_default.cmake:
```diff
@@ -81,9 +82,43 @@ include(${BOARD_DIR}/pre_dt_board.cmake OPTIONAL)
+
+if(NOT DEFINED SUB_COMPONENTS)
+  include(kernel)
+endif()
```
Aren't we being overly-generic here?

Isn't it much simpler and easier for people to understand this code if we did

if(STOP_AFTER_DT)
return()
endif()


if(STOP_AFTER_KCONF)
return()
endif()

tejlmand replied:
> Aren't we being overly-generic here?
> 
> Isn't it much simpler and easier for people to understand this code if we did

For a maintainer of this code, I partly agree.
But not for the users of the tool.

I don't expect users of this tool to have internal knowledge of the order in which Zephyr CMake modules are loaded.

So if a user or twister needs to have DTS and kconfig loaded they should not need to know which part is loaded first, they can simply say, *I want dts and kconfig* by invoking:
```
$ cmake -DBOARD=<board> -S samples/hello_world -DMODULES="kconfig;dts" -P cmake/package_tooling.cmake
```

and the script will ensure that both dts and kconfig are loaded.

With your proposal, a user / tool will need to consider, *I want both dts and kconfig, is dts executed loaded first, or is kconfig ?*, so do I need to do: `-DSTOP_AFTER_KCONF=true` or `-DSTOP_AFTER_DT=true`.

tejlmand replied:
not to mention, the current proposal ensures the Zephyr CMake module names are used, and not a derived setting name.

If user makes a typo or otherwise writes an invalid module name, for example: `-DMODULES=dtts` an error like this is printed:
```
Error: Subcomponent dtts not default module for Zephyr CMake build system.
         Please choose one or more valid components: <whole-list-of-valid-modules>"
```

whereas with your proposal, users need to know the relation between the module name and corresponding CMake variable, like: `dts` module corresponds to `STOP_AFTER_DT`.

And if writing `-DSTOP_AFTER_DTS=true` instead, then everything is processed and no error message is returned to the user.

SebastianBoe replied:
IMHO it is reasonable to expect that anyone that wants to halt execution after both DT and Kconfig have executed knows that DT executes before Kconfig.

IMHO it is also acceptable that obscure errors happen if the user uses the wrong module name.

This niche feature just isn't worth all this complexity we are adding.

tejlmand replied:
> This niche feature just isn't worth all this complexity we are adding.

we are adding Zephyr CMake modules to a list, and then looping that list until all modules requested by user has been loaded.

How can adding to a list and looping that list be considered complex ?

SebastianBoe replied:
To me it is very clear that adding two return statements and two if statements will be much simpler
for the user to understand than the patch we have now. I'm not sure how to argue the point.
It's just a matter of the amount of code, logic, and abstractions, needed to get the job done.

That being said, if you insist on this approach then we should at least reduce the compexity as much as we can.

e.g. we can create a function:

```
# write to sorted_sub_list_output, a sublist of sorted_list_input starting from the beginning of the list that contains
# all items in list_input
find_sublist_with_all(sorted_list_input list_input sorted_sub_list_output)
```

to remove some of the noise from the flow of the program.

tejlmand replied:
> To me it is very clear that adding two return statements and two if statements will be much simpler for the user to understand than the patch we have now.

Aah, so you think we should support only DTS and Kconfig.
I see this as a general way to allow us to run just run up to a certain point.
Other use-cases could be `zephyr_modules` or toolchain handling.

For example, we might be able to clean up this in future (outside the scope of this PR):
https://github.com/zephyrproject-rtos/zephyr/blob/8751a0feb0388698698197f08fbf5399917650c2/scripts/pylib/twister/twisterlib.py#L3029-L3030

in that case, you would need to add a CMake variable and check at each possible return point someone might use, not just two if and two return statements.

SebastianBoe replied:
complexity is now acceptable after the cleanup.

SebastianBoe commented on cmake/package_tooling.cmake:
```diff
@@ -0,0 +1,118 @@
+
+function(set_target_properties)
+  # This silence the error: 'set_target_properties command is not scriptable'
+endfunction()
```
Why are we invoking CMake in script mode?

tejlmand replied:
because we don't want to start creating an actual build system with targets, etc.

Also, if this was a real project, then you would execute:
```
$ cmake -DBOARD=<board> -S <path-to>/package_tooling
```

so how will you specify the actual sample for which you want to run this tool ?
You cannot do:
```
$ cmake -DBOARD=<board> -S <path-to>/package_tooling -S samples/hello_world
```

I try to keep the invocation aligned so that `west build --some-new-arg` and `cmake ...` are closely aligned to creating the build system.

For example, generating build system:
```
$ cmake -DBOARD=<board> -B build -S samples/hello_world
``` 

Running the tool:
```
$ cmake -DBOARD=<board> -B build -S samples/hello_world -DMODULES=dts -P <ZEPHYR_BASE>/cmake/package_tooling.cmake
```

This means, everything before ` -DMODULES` is identical when creating a build system as well as running the tool.

SebastianBoe replied:
Understood.

I feel like this is a lot of code (which will need to be maintained as more functions that are not supported in script mode are added), just for the optimization of omitting the generation step.

But it doesn't increase the complexity of the normal build, so this is acceptable.

SebastianBoe replied:
"so how will you specify the actual sample for which you want to run this tool ?"

The alternative would be for the tool to not exist and for users to invoke CMake normally.

SebastianBoe commented on cmake/modules/zephyr_default.cmake:
```diff
@@ -81,9 +82,37 @@ include(${BOARD_DIR}/pre_dt_board.cmake OPTIONAL)
+endforeach()
+
+if(NOT DEFINED SUB_COMPONENTS)
+  include(kernel)
```
shouldn't we unconditionally include kernel?

If SUB_COMPONENTS is defined the program will have returned before now.

tejlmand replied:
good point.

after the cleanup, this if can be removed.

will do.

tejlmand replied:
done

tejlmand commented:
Re-based, @mbolivar-nordic @SebastianBoe please re-approve.
